### PR TITLE
Fix query fail when there are timestamp or time columns after generated column

### LINKED ISSUE #7455: Duration filter push down cause query fail
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table if not exists test.t(a time(4), I int);
insert into test.t values('-700:10:10.123456', 1);
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
insert into test.t select * from test.t;
alter table test.t set tiflash replica 1;
explain analyze select * from test.t where a = '700:11:11.123500';
```

### 2. What did you expect to see? (Required)

```
mysql> explain analyze select * from test.t where a = '700:11:11.123500';
ERROR 1105 (HY000): other error for mpp stream: From MPP<query:<query_ts:1683783901339381008, local_query_id:3, server_id:1814245, start_ts:441393847025532929>,task_id:1>: Code: 49, e.displayText() = DB::Exception: Check schema[i].type->equals(*storage_schema[i].type) failed, schema[i].name = table_scan_0, schema[i].type->getName() = Nullable(MyDuration(4)), storage_schema[i].name = a, storage_schema[i].type->getName() = Nullable(Int64), e.what() = DB::Exception,
```

### 3. What did you see instead (Required)

success

### 4. What is your TiFlash version? (Required)

master

### LINKED ISSUE #7468: Query fail when there are timestamp or time columns after generated column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
CREATE TABLE test.`IDT_26539` (`COL102` float DEFAULT NULL, `COL103` float DEFAULT NULL, `COL1` float GENERATED ALWAYS AS ((`COL102` DIV 10)) VIRTUAL, `COL2` varchar(20) COLLATE utf8mb4_bin DEFAULT NULL, `COL4` time DEFAULT NULL, `COL3` bigint DEFAULT NULL, `COL5` float DEFAULT NULL, KEY `UK_COL1` (`COL1`) /*!80000 INVISIBLE */) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
insert into test.IDT_26539 (COL102, COL103, COL2, COL4, COL3, COL5) values (NULL, NULL, NULL, NULL, NULL, NULL);
insert into test.IDT_26539 (COL102, COL103, COL2, COL4, COL3, COL5) values (NULL, NULL, 'r2Ic', NULL, NULL, NULL);
alter table test.IDT_26539 set tiflash replica 1;
set tidb_isolation_read_engines='tiflash'; select * from test.IDT_26539 where col2 = 'r2Ic';
```

### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

```
mysql> set tidb_isolation_read_engines='tiflash'; select * from test.IDT_26539 where col2 = 'r2Ic';
Query OK, 0 rows affected (0.00 sec)

ERROR 1105 (HY000): Not found column generated_column_2 in block. There are only columns: col102, col103, col2, col4, col3, col5, FunctionConvertDurationFromNanos(col4, 0_Int64)_collator_0
```

### 4. What is your TiFlash version? (Required)

master

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #7468, close #7455

Problem Summary:

### What is changed and how it works?

1. handle generated column before handling timezone/duration cast
2. handle column type changed after duration type filter pushed down
3. refine some code in StorageDisaggregated

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/StorageDeltaMerge.cpp [hunk 1]
```diff
@@ -839,7 +839,22 @@ DM::PushDownFilterPtr StorageDeltaMerge::buildPushDownFilter(const RSOperatorPtr
...
+            const auto & source_columns = analyzer->getCurrentInputColumns();
+            for (size_t i = 0; i < table_scan_column_info.size(); ++i)
+            {
+                if (table_scan_column_info[i].hasGeneratedColumnFlag())
+                    continue;
+                auto it = columns_to_read_map.at(table_scan_column_info[i].id);
+                RUNTIME_CHECK(it.name == source_columns[i].name);
+                columns_after_cast->push_back(it);
+                columns_after_cast->back().type = source_columns[i].type;
+            }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/StorageDeltaMerge.cpp:855]: 1. Is `columns_after_cast` a better name than `source_columns`?
2. is `column` a better name than `it`?
3. Add necessary message: column id, left column name, right column name by `RUNTIME_CHECK_MSG`

Comment (Lloyd-Pottiger) [reply dbms/src/Storages/StorageDeltaMerge.cpp]: fixed

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 2]
```diff
@@ -431,11 +441,20 @@ void DAGStorageInterpreter::executeImpl(DAGPipeline & pipeline)
...
+    NamesAndTypes source_columns;
+    source_columns.reserve(table_scan.getColumnSize());
+    const auto table_scan_output_header = pipeline.firstStream()->getHeader();
+    for (const auto & col : table_scan_output_header)
+        source_columns.emplace_back(col.name, col.type);
+    analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
+    /// If there is no local stream, there is no need to execute cast and push down filter, return directly.
+    /// But we should make sure that the analyzer is initialized before return.
+    if (remote_read_streams_start_index == 0)
+        return;
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:455]: But `recordProfileStreams` is skipped?

Comment (Lloyd-Pottiger) [reply dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp]: fixed

**Review by JaySon-Huang — APPROVED** [review]
lgtm

File: dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp [hunk 3]
```diff
@@ -431,11 +441,23 @@ void DAGStorageInterpreter::executeImpl(DAGPipeline & pipeline)
...
     executeGeneratedColumnPlaceholder(remote_read_streams_start_index, generated_column_infos, log, pipeline);
+    NamesAndTypes source_columns;
+    source_columns.reserve(table_scan.getColumnSize());
+    const auto table_scan_output_header = pipeline.firstStream()->getHeader();
+    for (const auto & col : table_scan_output_header)
+        source_columns.emplace_back(col.name, col.type);
+    analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
+    /// If there is no local stream, there is no need to execute cast and push down filter, return directly.
+    /// But we should make sure that the analyzer is initialized before return.
+    if (remote_read_streams_start_index == 0)
```
Comment (SeaRise) [inline dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:454]: The recordProfileStreams of the pushdown filter are also skipped. It seems that this early return is actually unnecessary?
L467

Comment (Lloyd-Pottiger) [reply dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp:454]:
```cpp
void executePushedDownFilter(
    size_t remote_read_streams_start_index,
    const FilterConditions & filter_conditions,
    DAGExpressionAnalyzer & analyzer,
    LoggerPtr log,
    DAGPipeline & pipeline)
{
    auto [before_where, filter_column_name, project_after_where] = ::DB::buildPushDownFilter(filter_conditions.conditions, analyzer);

    assert(remote_read_streams_start_index <= pipeline.streams.size());
    // for remote read, filter had been pushed down, don't need to execute again.
    for (size_t i = 0; i < remote_read_streams_start_index; ++i)
    {
        auto & stream = pipeline.streams[i];
        stream = std::make_shared<FilterBlockInputStream>(stream, before_where, filter_column_name, log->identifier());
        stream->setExtraInfo("push down filter");
        // after filter, do project action to keep the schema of local streams and remote streams the same.
        stream = std::make_shared<ExpressionBlockInputStream>(stream, project_after_where, log->identifier());
        stream->setExtraInfo("projection after push down filter");
    }
}
```

When remote_read_streams_start_index == 0, there is no pushdown filter.

When remote_read_streams_start_index == 0, the schema of tablescan_output_header has been changed(MyTime --> MyDuration), so executeCastAfterTableScan will report error.

**Pull request merged**